packages/react-hooks/src/useConst.ts

```diff
@@ -0,0 +1,29 @@
+import * as React from 'react';
+
+/**
+ * Hook to initialize and return a constant value. Unlike `React.useMemo`, this is guaranteed
+ * to always return the same value (and only ever call the initializer function once).
+ * This is similar to setting a private member in a class constructor.
+ *
+ * If the value should ever change based on dependencies, use `React.useMemo` instead.
+ *
+ * @param initializer - Function to get the initial value. Similar to `useState`, this function
+ * will only be called the first time the hook is used, and updates won't be respected.
+ * @returns The value. The identity of this value will always be the same.
+ *
+ * {@docCategory Hooks}
+ */
+export function useConst<T>(initializer: () => T): T {
```

Comment:
I think we should be as close as possible to existing React Hook APIs. It feels weird that I can initialize something as simple as `useConst` only with a lambda function, and not a normal value, unlike `useState`.

I acknowledge that we can't store function components without boxing them, but it would be consistent with `useState`, and we don't have to create unnecessary lambda functions on every render for each instance of this very simple hook.

Reply:
The change to require a function initializer was based on @ThomasMichon's feedback over chat, but I'm open to other input. The new `useConstCallback` hook would help with the concern about storing functions (though the semantics aren't ideal for storing components).

As context for anyone else reading this, here's Thomas's comment about storing function components:

> I'd do away with `T | () => T` because what if the state you want to store is a function? I remember some teammates getting confused by stuff like:
> 
> ```
> const [state, setState] = useState();
> useEffect(() => {
>   import('./FooComponent').then((component) => {
>     setState(component);
>   });
> }, []);
> ```
> Basically, FooComponent is a function, so useState thinks you just passed it a reducer. The 'fix' is `setState(() => component)` if your value is itself a function.

Reply:
Agreee with @soettl , we should take both. Have useConstCallback for the other scenario  :)

Reply:
Actually in thinking about this, I can't intuitively say why I would use `useConst` or `useConstCallback`.

Why not just use `useState`? Calling a function to define each variable seems really unneeded.

```tsx
const MyComponent = props => {
   const internalState = React.useState({
     const1: val,
     callback1: () => etc
   });
}
```